### PR TITLE
CIRC-199 Alternate loan period with pending hold

### PR DESCRIPTION
**Pull request by rladdusaw**
Resolves [CIRC-199](https://issues.folio.org/browse/CIRC-199)

## Purpose
This PR introduces two changes in the initial due date applied by a loan policy. If the second request in a queue is both a hold and has a status of OPEN_NOT_YET_FILLED, then the alternateCheckoutLoanPeriod from the holds on a loan policy will be applied. This PR will enact the desired behavior in both Rolling and Fixed due date schedules.

## Approach
Significant changes to the determineStrategy method and DueDateStrategy 
classes were needed to obtain the information needed to perform the
necessary logic.

#### TODOS and Open Questions
- Open Questions
  - [ ] FixedDueDateSchedules use a final list of schedules. If the ability to add schedules were implemented, some code could be cleaned up. Should this be done, or is it more important to preserve the finality of this property
  - [ ] Is "alternateDueDateSchedule" an appropriate identifier for the [schedule](https://github.com/folio-org/mod-circulation/blob/05e9037242d0bd0c63ac8e7a957ca80bb4b15a42/src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java#L290) created

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by marcjohnson-kint** [comment]
>  FixedDueDateSchedules use a final list of schedules. If the ability to add schedules were implemented, some code could be cleaned up. Should this be done, or is it more important to preserve the finality of this property

The intent behind class design in mod-circulation is to prefer immutability. This hasn't been achieved as well I might've likely. If it is causing issues with design, then please change it.

> Is "alternateDueDateSchedule" an appropriate identifier for the schedule created

Hmm, This is where the loan policy design is likely creaking. At one point it was intended to represent the policy in use in a particular circumstance.

Now that those scenarios have expanded, it is likely less and less appropriate for this to be modelled this way.

I increasingly want to move loan policy towards being a data class / structure and move the various different contexts to services or workflows, because it is trying to do way too much.

**Comment by marcjohnson-kint** [comment]
@rladdusaw I think this pull request might need closing and re-submitting, as Jenkins is not able to pick it up and run a build.

**Comment by marcjohnson-kint** [comment]
> @rladdusaw I think this pull request might need closing and re-submitting, as Jenkins is not able to pick it up and run a build.

Scratch that, I closed and re-opened it and the build is now running.

File: src/test/java/api/support/builders/LoanPolicyBuilder.java [hunk 1]
```diff
@@ -591,6 +626,30 @@ public LoanPolicyBuilder withRecallsRecallReturnInterval(Period recallsRecallRet
       this.alternateFixedDueDateScheduleId,
       this.renewable,
       this.recallsMinimumGuaranteedLoanPeriod,
-      recallsRecallReturnInterval.asJson());
+      recallsRecallReturnInterval.asJson(),
+      this.alternateCheckoutLoanPeriod);
+  }
+
+  public LoanPolicyBuilder withAlternateCheckoutLoanPeriod(Period alternateCheckoutLoanPeriod) {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/LoanPolicyBuilder.java]: I think it is important that this refers to the situation when it applies, in this case it is hold requests (admittedly, my suggestion has compromises due to length)

```suggestion
  public LoanPolicyBuilder withHoldRequestAlternateLoanPeriod(Period alternateLoanPeriod) {
```

File: src/test/java/org/folio/circulation/domain/policy/FixedLoanPolicyRenewalDueDateCalculationTests.java [hunk 2]
```diff
@@ -223,6 +233,64 @@ public void shouldUseAlternateScheduleWhenAvailable() {
...
+    String key = "alternateCheckoutLoanPeriod";
+    DateTime expectedDueDate = alternateCheckoutLoanPeriod.addTo(
+        systemTime,
+        () -> errorForLoanPeriod(format("the \"%s\" is not recognized", key)),
+        interval -> errorForLoanPeriod(format("the interval \"%s\" in \"%s\" is not recognized", interval, key)),
+        dur -> errorForLoanPeriod(format("the duration \"%s\" in \"%s\" is invalid", dur, key)))
+          .value();
+    Long result = initialDueDateResult.value().getMillis();
+    Long expected = expectedDueDate.getMillis();
+    assertThat(result.doubleValue(), closeTo(expected.doubleValue(), 60000));
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/domain/policy/FixedLoanPolicyRenewalDueDateCalculationTests.java]: What does this refer to?

Personally, I would find it more readable if the assertions referred to the date and time that was expected in a more human friendly format (I believe the other tests do this)

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java [hunk 3]
```diff
@@ -33,18 +32,34 @@ static FixedDueDateSchedules from(JsonObject representation) {
...
+  public List<JsonObject> getSchedules() {
+    return schedules;
+  }
+
   public Optional<DateTime> findDueDateFor(DateTime date) {
     return findScheduleFor(date)
       .map(this::getDueDate);
   }
 
+  public Optional<DateTime> findEarliestDueDateFor(DateTime date) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: I don't really understand what this method is doing.

What is it intended to do?

Can you help me understand what scenario it matches up with in the story?

**Comment by jsavell** [comment]
Closing in favor of #356

**Pull request closed without merging**